src/main/java/dp/xlsx/V4File.java

```diff
@@ -25,57 +27,69 @@
  */
 class V4File {
 
-    private final List<String[]> data; // v4 CSV rows
-    private final int headerOffset;
-    private final Group headerGroup;
-    private Set<String> uniqueTimeValues;
+    private final Collection<Group> groupData;
+    private final Set<String> uniqueTimeValues;
+    private Group headerGroup;
 
     V4File(final InputStream inputStream) throws IOException {
-        try (final CSVReader reader = new CSVReader(new InputStreamReader(inputStream))) {
-            data = reader.readAll();
-            if (data.size() <= 1) {
-                throw new IOException("Two or more csv rows are need to generate a XLSX file");
-            }
-        }
 
-        final String[] header = data.get(0);
-        final String v4Code = header[0];
+        final Map<Group, Group> groups = new HashMap<>();
 
-        headerOffset = Integer.parseInt(v4Code.split("V4_")[1]) + 1;
-        uniqueTimeValues = new HashSet<>();
-        headerGroup = new Group(header, headerOffset);
-    }
+        try (final InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
+             final BufferedReader bufferedReader = new BufferedReader(inputStreamReader)) {
 
-    /**
-     * Group the v4 file by all dimension except for the time dimension
-     *
-     * @return A list of all groups within the v4 file
-     */
-    List<Group> groupData() {
+            uniqueTimeValues = new HashSet<>();
 
-        data.remove(0); // remove header
+            String line;
+            int headerOffset = 0;
 
-        final Map<Group, Group> groups = new HashMap<>();
+            CsvParserSettings settings = new CsvParserSettings();
+            settings.setEmptyValue("");
+            settings.setNullValue("");
+            CsvParser parser = new CsvParser(settings);
+
+            if ((line = bufferedReader.readLine()) != null) {
+
+                final String[] header = parser.parseLine(line);
+                final String v4Code = header[0];
 
-        data.stream().forEach(row -> {
+                headerOffset = Integer.parseInt(v4Code.split("V4_")[1]) + 1;
+                headerGroup = new Group(header, headerOffset);
+            }
+
+            while ((line = bufferedReader.readLine()) != null) {
 
-            final Group group = new Group(row, headerOffset);
-            final String timeValue = row[headerOffset + 1];
-            final String observation = row[0];
+                final String[] row = parser.parseLine(line);
```

Comment:
This can return a null string array

Reply:
I have skipped the line if its parse result is null